server/lib/src/idm/oauth2.rs

```diff
@@ -2701,12 +2835,49 @@ fn validate_scopes(req_scopes: &BTreeSet<String>) -> Result<(), Oauth2Error> {
     Ok(())
 }
 
+/// device code is a random bucket of bytes used in the device flow
+#[inline]
+fn gen_device_code() -> Result<[u8; 16], Oauth2Error> {
+    let mut rng = rand::thread_rng();
+    let mut result = [0u8; 16];
+    if let Err(err) = rng.try_fill(&mut result) {
+        error!("Failed to generate device code! {:?}", err);
+        return Err(Oauth2Error::ServerError(OperationError::Backend));
+    }
+    Ok(result)
+}
+
+#[inline]
+/// Returns (xxx-yyy-zzz, digits) where one's the human-facing code, the other is what we store in the DB.
+fn gen_user_code() -> (String, u32) {
+    use rand::Rng;
+    let mut rng = rand::thread_rng();
+    let num: u32 = rng.gen_range(0..=999999999);
+    let result = format!("{:09}", num);
```

Comment:
There is probably a modulo or subtract trick you could do here ...

Reply:
All I'm doing is zero-padding it for display purposes, why bother with tricks?

Reply:
I meant so that you don't need to double format.

Reply:
Like you could do format!("{:03}-{:03}-{:03}", num /  1000000,  (num / 1000) % 1000,  num % 1000  );

Reply:
yes, I'm micro-optimising.